File: single-threaded-cpu.cc
// Single-Threaded CPU
#define ALL(x) (x).begin(), (x).end()
#define FOR(i, a, b) for (remove_cv<remove_reference<decltype(b)>::type>::type i = (a); i < (b); i++)
#define REP(i, n) FOR(i, 0, n)

struct Node {
  int bgn, dur, idx;
  bool operator>(const Node &r) const {
    return tie(dur, idx) > tie(r.dur, r.idx);
  }
};

class Solution {
public:
  vector<int> getOrder(vector<vector<int>>& tasks) {
    vector<Node> a;
    REP(i, tasks.size())
      a.push_back(Node{tasks[i][0], tasks[i][1], (int)i});
    sort(ALL(a), [](const Node &l, const Node &r) {
      return l.bgn != r.bgn ? l.bgn < r.bgn : l.idx < r.idx;
    });
    priority_queue<Node, vector<Node>, greater<Node>> b;
    int n = a.size(), i = 0;
    long tick = 0;
    vector<int> ans;
    while (i < n || b.size()) {
      if (i < n && a[i].bgn <= tick)
        b.push(a[i++]);
      else if (!b.empty()) {
        ans.push_back(b.top().idx);
        tick += b.top().dur;
        b.pop();
      } else
        tick = a[i].bgn;
    }
    return ans;
  }
};

File: single-threaded-cpu.rs
impl Solution {
    pub fn get_order(tasks: Vec<Vec<i32>>) -> Vec<i32> {
        use std::collections::BinaryHeap;
        use std::cmp::Reverse;
        let n = tasks.len();
        let mut tasks: Vec<(i32, i32, usize)> = tasks
            .into_iter().enumerate().map(|(i,a)| (a[0], a[1], i)).collect();
        tasks.sort_by_key(|v| v.0);

        let mut q: BinaryHeap<Reverse<(i32, usize)>> = BinaryHeap::new();
        let mut ans = Vec::new();
        let mut tick = 0;
        let mut i = 0;
        while ans.len() != n {
            if i < n && tasks[i].0 <= tick {
                q.push(Reverse((tasks[i].1, tasks[i].2)));
                i += 1;
            } else if !q.is_empty() {
                let x = q.pop().unwrap().0;
                ans.push(x.1 as i32);
                tick += x.0;
            } else {
                tick = tasks[i].0;
            }
        }
        ans
    }
}
